### PR TITLE
Fix port validation error on specifying tcp/udp or range of ports.

### LINKED ISSUE #13809: Port validation error on specifying tcp/udp or range ports
### What Happened?

`$ minikube start` fails on specifying tcp, udp or range ports option.
e.g. `127.0.0.1:8080:8080/tcp`, `127.0.0.1:8080:8080/udp`, `127.0.0.1:8080-8081:8080-8081`.

This problem appears to have been introduced on minikube v1.24.0 (#12233).

#### 127.0.0.1:8080:8080/tcp

```sh
$ minikube start --driver=docker --ports=127.0.0.1:8080:8080/tcp
😄  minikube v1.25.2 on Darwin 12.2.1 (arm64)
✨  Using the docker driver based on user configuration

❌  Exiting due to MK_USAGE: Sorry, one of the ports provided with --ports flag is not valid [127.0.0.1:8080:8080/tcp]
```

#### 127.0.0.1:8080:8080/udp

```sh
$ minikube start --driver=docker --ports=127.0.0.1:8080:8080/udp
😄  minikube v1.25.2 on Darwin 12.2.1 (arm64)
✨  Using the docker driver based on user configuration

❌  Exiting due to MK_USAGE: Sorry, one of the ports provided with --ports flag is not valid [127.0.0.1:8080:8080/udp]
```

#### 127.0.0.1:8080-8081:8080-8081

```sh
$ minikube start --ports=127.0.0.1:8080-8081:8080-8081
😄  minikube v1.25.2 on Darwin 12.2.1 (arm64)
✨  Automatically selected the docker driver

❌  Exiting due to MK_USAGE: Sorry, one of the ports provided with --ports flag is not valid [127.0.0.1:8080-8081:8080-8081]
```

#### 127.0.0.1:8080:8080/tcp

```sh
$ minikube start --driver=docker --ports=127.0.0.1:8080:8080/tcp
😄  minikube v1.23.2 on Darwin 12.2.1 (arm64)
✨  Using the docker driver based on user configuration
👍  Starting control plane node minikube in cluster minikube
🚜  Pulling base image ...
🔥  Creating docker container (CPUs=2, Memory=4000MB) ...
🐳  Preparing Kubernetes v1.22.2 on Docker 20.10.8 ...
    ▪ Generating certificates and keys ...
    ▪ Booting up control plane ...
    ▪ Configuring RBAC rules ...
🔎  Verifying Kubernetes components...
    ▪ Using image gcr.io/k8s-minikube/storage-provisioner:v5
🌟  Enabled addons: storage-provisioner, default-storageclass
🏄  Done! kubectl is now configured to use "minikube" cluster and "default" namespace by default

$ docker ps -a
CONTAINER ID   IMAGE                                 COMMAND                  CREATED          STATUS          PORTS                                                                                                                                                            NAMES
0790c0fb3393   gcr.io/k8s-minikube/kicbase:v0.0.27   "/usr/local/bin/entr…"   34 seconds ago   Up 33 seconds   127.0.0.1:8080->8080/tcp, 127.0.0.1:62515->22/tcp, 127.0.0.1:62511->2376/tcp, 127.0.0.1:62513->5000/tcp, 127.0.0.1:62514->8443/tcp, 127.0.0.1:62512->32443/tcp   minikube
```

#### 127.0.0.1:8080:8080/udp

```sh
$ minikube start --driver=docker --ports=127.0.0.1:8080:8080/udp
😄  minikube v1.23.2 on Darwin 12.2.1 (arm64)
✨  Using the docker driver based on user configuration
👍  Starting control plane node minikube in cluster minikube
🚜  Pulling base image ...
🔥  Creating docker container (CPUs=2, Memory=4000MB) ...
🐳  Preparing Kubernetes v1.22.2 on Docker 20.10.8 ...
    ▪ Generating certificates and keys ...
    ▪ Booting up control plane ...
    ▪ Configuring RBAC rules ...
🔎  Verifying Kubernetes components...
    ▪ Using image gcr.io/k8s-minikube/storage-provisioner:v5
🌟  Enabled addons: storage-provisioner, default-storageclass
🏄  Done! kubectl is now configured to use "minikube" cluster and "default" namespace by default

$ docker ps -a
CONTAINER ID   IMAGE                                 COMMAND                  CREATED          STATUS          PORTS                                                                                                                                                            NAMES
36324c4cefa2   gcr.io/k8s-minikube/kicbase:v0.0.27   "/usr/local/bin/entr…"   33 seconds ago   Up 32 seconds   127.0.0.1:8080->8080/udp, 127.0.0.1:61975->22/tcp, 127.0.0.1:61976->2376/tcp, 127.0.0.1:61978->5000/tcp, 127.0.0.1:61979->8443/tcp, 127.0.0.1:61977->32443/tcp   minikube
```

#### 127.0.0.1:8080-8081:8080-8081

```sh
$ minikube start --driver=docker --ports=127.0.0.1:8080-8081:8080-8081
😄  minikube v1.23.2 on Darwin 12.2.1 (arm64)
✨  Using the docker driver based on user configuration
👍  Starting control plane node minikube in cluster minikube
🚜  Pulling base image ...
🔥  Creating docker container (CPUs=2, Memory=4000MB) ...
🐳  Preparing Kubernetes v1.22.2 on Docker 20.10.8 ...
    ▪ Generating certificates and keys ...
    ▪ Booting up control plane ...
    ▪ Configuring RBAC rules ...
🔎  Verifying Kubernetes components...
    ▪ Using image gcr.io/k8s-minikube/storage-provisioner:v5
🌟  Enabled addons: storage-provisioner, default-storageclass
🏄  Done! kubectl is now configured to use "minikube" cluster and "default" namespace by default

$ docker ps -a
CONTAINER ID   IMAGE                                 COMMAND                  CREATED              STATUS              PORTS                                                                                                                                                                      NAMES
69d5e11105ed   gcr.io/k8s-minikube/kicbase:v0.0.27   "/usr/local/bin/entr…"   About a minute ago   Up About a minute   127.0.0.1:8080-8081->8080-8081/tcp, 127.0.0.1:61085->22/tcp, 127.0.0.1:61087->2376/tcp, 127.0.0.1:61090->5000/tcp, 127.0.0.1:61091->8443/tcp, 127.0.0.1:61089->32443/tcp   minikube
```

### Attach the log file

No log file because minikube can't start with these port option.

### Operating System

macOS (Default)

### Driver

Docker

### PR DESCRIPTION
**Pull request by tyabu12**
Fixed #13809

### Tests

```sh
$ ./out/minikube start --driver=docker --ports=127.0.0.1:8080:8080/tcp 
😄  minikube v1.25.2 on Darwin 12.2.1 (arm64)
✨  Using the docker driver based on user configuration
👍  Starting control plane node minikube in cluster minikube
🚜  Pulling base image ...
🔥  Creating docker container (CPUs=2, Memory=4000MB) ...
🐳  Preparing Kubernetes v1.23.3 on Docker 20.10.12 ...
    ▪ Generating certificates and keys ...
    ▪ Booting up control plane ...
    ▪ Configuring RBAC rules ...
🔎  Verifying Kubernetes components...
    ▪ Using image gcr.io/k8s-minikube/storage-provisioner:v5
🌟  Enabled addons: storage-provisioner, default-storageclass
🏄  Done! kubectl is now configured to use "minikube" cluster and "default" namespace by default

$ docker ps 
CONTAINER ID   IMAGE                                 COMMAND                  CREATED          STATUS          PORTS                                                                                                                                                            NAMES
71a90cb64130   gcr.io/k8s-minikube/kicbase:v0.0.30   "/usr/local/bin/entr…"   23 seconds ago   Up 22 seconds   127.0.0.1:8080->8080/tcp, 127.0.0.1:51870->22/tcp, 127.0.0.1:51871->2376/tcp, 127.0.0.1:51873->5000/tcp, 127.0.0.1:51869->8443/tcp, 127.0.0.1:51872->32443/tcp   minikube
```

```sh
$ ./out/minikube start --driver=docker --ports=127.0.0.1:8080:8080/udp
😄  minikube v1.25.2 on Darwin 12.2.1 (arm64)
✨  Using the docker driver based on user configuration
👍  Starting control plane node minikube in cluster minikube
🚜  Pulling base image ...
🔥  Creating docker container (CPUs=2, Memory=4000MB) ...
🐳  Preparing Kubernetes v1.23.3 on Docker 20.10.12 ...
    ▪ Generating certificates and keys ...
    ▪ Booting up control plane ...
    ▪ Configuring RBAC rules ...
🔎  Verifying Kubernetes components...
    ▪ Using image gcr.io/k8s-minikube/storage-provisioner:v5
🌟  Enabled addons: storage-provisioner, default-storageclass
🏄  Done! kubectl is now configured to use "minikube" cluster and "default" namespace by default

$ docker ps
CONTAINER ID   IMAGE                                 COMMAND                  CREATED          STATUS          PORTS                                                                                                                                                            NAMES
d5de1e011cb3   gcr.io/k8s-minikube/kicbase:v0.0.30   "/usr/local/bin/entr…"   18 seconds ago   Up 17 seconds   127.0.0.1:8080->8080/udp, 127.0.0.1:52360->22/tcp, 127.0.0.1:52356->2376/tcp, 127.0.0.1:52358->5000/tcp, 127.0.0.1:52359->8443/tcp, 127.0.0.1:52357->32443/tcp   minikube
```

```sh
$ ./out/minikube start --driver=docker --ports=127.0.0.1:8080-8081:8080-8081
😄  minikube v1.25.2 on Darwin 12.2.1 (arm64)
✨  Using the docker driver based on user configuration
👍  Starting control plane node minikube in cluster minikube
🚜  Pulling base image ...
🔥  Creating docker container (CPUs=2, Memory=4000MB) ...
🐳  Preparing Kubernetes v1.23.3 on Docker 20.10.12 ...
    ▪ Generating certificates and keys ...
    ▪ Booting up control plane ...
    ▪ Configuring RBAC rules ...
🔎  Verifying Kubernetes components...
    ▪ Using image gcr.io/k8s-minikube/storage-provisioner:v5
🌟  Enabled addons: storage-provisioner, default-storageclass
🏄  Done! kubectl is now configured to use "minikube" cluster and "default" namespace by default

$ docker ps
CONTAINER ID   IMAGE                                 COMMAND                  CREATED          STATUS          PORTS                                                                                                                                                                      NAMES
be4e51639805   gcr.io/k8s-minikube/kicbase:v0.0.30   "/usr/local/bin/entr…"   18 seconds ago   Up 18 seconds   127.0.0.1:8080-8081->8080-8081/tcp, 127.0.0.1:53079->22/tcp, 127.0.0.1:53080->2376/tcp, 127.0.0.1:53077->5000/tcp, 127.0.0.1:53078->8443/tcp, 127.0.0.1:53076->32443/tcp   minikube
```

```sh
$ ./out/minikube start --driver=docker --ports=127.0.0.1:8080-8081:8080-8081/udp
😄  minikube v1.25.2 on Darwin 12.2.1 (arm64)
✨  Using the docker driver based on user configuration
👍  Starting control plane node minikube in cluster minikube
🚜  Pulling base image ...
🔥  Creating docker container (CPUs=2, Memory=4000MB) ...
🐳  Preparing Kubernetes v1.23.3 on Docker 20.10.12 ...
    ▪ Generating certificates and keys ...
    ▪ Booting up control plane ...
    ▪ Configuring RBAC rules ...
🔎  Verifying Kubernetes components...
    ▪ Using image gcr.io/k8s-minikube/storage-provisioner:v5
🌟  Enabled addons: storage-provisioner, default-storageclass
🏄  Done! kubectl is now configured to use "minikube" cluster and "default" namespace by default

$ docker ps
CONTAINER ID   IMAGE                                 COMMAND                  CREATED          STATUS          PORTS                                                                                                                                                                      NAMES
355278e5ca5a   gcr.io/k8s-minikube/kicbase:v0.0.30   "/usr/local/bin/entr…"   17 seconds ago   Up 16 seconds   127.0.0.1:8080-8081->8080-8081/udp, 127.0.0.1:53570->22/tcp, 127.0.0.1:53571->2376/tcp, 127.0.0.1:53568->5000/tcp, 127.0.0.1:53569->8443/tcp, 127.0.0.1:53567->32443/tcp   minikube
```

### CODE REVIEW

**Review by klaases — APPROVED** [review]

File: cmd/minikube/cmd/start.go [hunk 1]
```diff
@@ -1246,24 +1247,20 @@ func validateFlags(cmd *cobra.Command, drvName string) {
 
 // This function validates that the --ports are not below 1024 for the host and not outside range
```
[the remark this answers is not in the document]
Comment (tyabu12) [reply cmd/minikube/cmd/start.go]: @klaases 
Thank you for reviewing.
I updated the function comment at [513a757ca7b5cf3cf88be5494a897b091c5983d6] .

File: cmd/minikube/cmd/start_test.go [hunk 2]
```diff
@@ -462,40 +462,124 @@ func TestValidateRuntime(t *testing.T) {
 }
 
 func TestValidatePorts(t *testing.T) {
+	isMicrosoftWSL := detect.IsMicrosoftWSL()
 	type portTest struct {
+		isTarget bool
```
Comment (klaases) [inline cmd/minikube/cmd/start_test.go:467]: Hi @tyabu12, can you add a comment to help us how this works?

File: cmd/minikube/cmd/start.go [hunk 3]
```diff
@@ -1244,26 +1245,22 @@ func validateFlags(cmd *cobra.Command, drvName string) {
...
-		if len(parts) > 2 {
-			ip := parts[0]
-			if net.ParseIP(ip) == nil {
-				return errors.Errorf("Sorry, the IP address provided with --ports flag is invalid: %s", ip)
-			}
-			parts = parts[1:]
-		}
-		for i, port := range parts {
-			p, err := strconv.Atoi(port)
+	_, portBindingsMap, err := nat.ParsePortSpecs(ports)
```
Comment (klaases) [inline cmd/minikube/cmd/start.go:1250]: Hi @tyabu12, can you confirm that this third-party library function offers improved functionality over the original code?

File: cmd/minikube/cmd/start_test.go (hunk 2 above)
Comment (tyabu12) [reply cmd/minikube/cmd/start_test.go:469]: @klaases 
I added  a comment at [afb3956fdbde357e4baa0f8617bfd5a64bad6558] .

File: cmd/minikube/cmd/start.go (hunk 3 above)
Comment (tyabu12) [reply cmd/minikube/cmd/start.go:1250]: @klaases 
docker-cli use the third-party library function to parse publish ports option.
https://github.com/docker/cli/blob/master/cli/command/container/opts.go#L416

Strictly speaking, convertToStandardNotation function converts target and published format at L411 (e.g.  `target=10,published=20` -> `20:10/tcp`), but the format seems to be unfamiliar.

**Comment by tyabu12** [comment]
@klaases
Thanks for your review.
I added the commits as you have mentioned. Can you review again?

**Comment by sharifelgamal** [comment]
/ok-to-test

Let's run the tests one more time, the first run was a bit wonky.

**Review by sharifelgamal — APPROVED** [review]
Thank you for your contribution!

**Pull request merged**